pkg/backend/backend.go

```diff
@@ -222,3 +223,7 @@ func (c *backendClient) GetStackOutputs(ctx context.Context, name string) (resou
 	}
 	return res.Outputs, nil
 }
+
+func (c *backendClient) DownloadPlugin(ctx context.Context, plug workspace.PluginInfo) (io.ReadCloser, error) {
+	return nil, errors.New("downloading plugins at runtime not available when using local backend")
```

Comment:
Any particular reason to make this behavior different?

Reply:
CLI verbs that hit a backend take a "cloud URL" parameter which is what `httpstate.Backend` talks to. The local backend ignores that parameter entirely, since it's not hitting an HTTP endpoint, but we'd actually need to hit an HTTP endpoint here in order to download a plugin. Terminating with an error is consistent with how we deal with other "not implemented" aspects of the local backend, such as stack tags.

Reply:
> The local backend ignores that parameter entirely, since it's not hitting an HTTP endpoint, but we'd actually need to hit an HTTP endpoint here in order to download a plugin.

IIRC `pulumi plugin install` still works with the local backend, and this should be the moral equivalent of that, no?

Reply:
Yeah, it's an unfortunate artifact of evolution that we use the currently connected cloud to download plugins.  We probably need to move to a model where by default we just always download plugins from somewhere like `get.pulumi.com`.  We have work scheduled to do that, and we can clean this up then.

Reply:
That's true, it does, but in my mind we've been communicating the local backend as "we will not talk to pulumi.com at all on the critical path of your deployment", and once we start downloading plugins at runtime, we're straying from that a bit.

This wouldn't be hard to do from a code perspective, but we would have to juggle some interfaces around in `pkg/backend` since `DownloadPlugins` is an `httpbackend.Backend`-only feature. We could promote `DownloadPlugin` to the full `Backend` interface, but then we'd need to have a cloud URL every time we instantiate a backend, which may or may not be a problem.

Would you be OK if we tackled that in a follow-up PR?

Reply:
> We probably need to move to a model where by default we just always download plugins from somewhere like get.pulumi.com. We have work scheduled to do that, and we can clean this up then.

Yes - it makes sense to me that these two things should be done together.

Reply:
Can we add a note in the tracking issue?

Reply:
https://github.com/pulumi/pulumi-service/issues/3068#issuecomment-473393068